pkg/replica/hash.go

```diff
@@ -246,93 +243,141 @@ func (t *SnapshotHashJob) isSnapshotSilentlyCorrupted(checksum string) bool {
+	stat := fileInfo.Sys().(*syscall.Stat_t)
+	return time.Unix(int64(stat.Ctim.Sec), int64(stat.Ctim.Nsec)).String(), nil
 }
 
-func GetSnapshotHashInfoFromXattr(snapshotName string) (*SnapshotXattrHashInfo, error) {
-	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
-	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue)
+func GetSnapshotHashInfoFromChecksumFile(snapshotName string) (*SnapshotXattrHashInfo, error) {
+	dir, err := os.Getwd()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get working directory")
+	}
+
+	path := filepath.Join(dir, diskutil.GenerateSnapshotDiskChecksumName(diskutil.GenerateSnapshotDiskName(snapshotName)))
+
+	f, err := os.Open(path)
 	if err != nil {
 		return nil, err
 	}
+	defer f.Close()
 
-	index := bytes.IndexByte(xattrSnapshotHashValue, 0)
+	var info SnapshotXattrHashInfo
 
-	info := &SnapshotXattrHashInfo{}
-	if err := json.Unmarshal(xattrSnapshotHashValue[:index], info); err != nil {
+	if err := json.NewDecoder(f).Decode(&info); err != nil {
 		return nil, err
 	}
 
-	return info, nil
+	return &info, nil
 }
 
-func SetSnapshotHashInfoToXattr(snapshotName string, info *SnapshotXattrHashInfo) error {
-	xattrSnapshotHashValue, err := json.Marshal(&SnapshotXattrHashInfo{
+func SetSnapshotHashInfoToChecksumFile(snapshotName string, info *SnapshotXattrHashInfo) error {
+	dir, err := os.Getwd()
+	if err != nil {
+		return errors.Wrap(err, "cannot get working directory")
+	}
+
+	path := filepath.Join(dir, diskutil.GenerateSnapshotDiskChecksumName(diskutil.GenerateSnapshotDiskName(snapshotName)))
+
+	return encodeToFile(SnapshotXattrHashInfo{
 		Method:            defaultHashMethod,
 		Checksum:          info.Checksum,
-		ModTime:           info.ModTime,
+		ChangeTime:        info.ChangeTime,
 		LastHashedAt:      info.LastHashedAt,
 		SilentlyCorrupted: info.SilentlyCorrupted,
-	})
+	}, path)
+}
+
+func encodeToFile(obj interface{}, path string) (err error) {
+	tmpPath := fmt.Sprintf("%s.%s", path, tmpFileSuffix)
+
+	defer func() {
+		var rollbackErr error
+		if err != nil {
+			if _, err := os.Stat(tmpPath); err == nil {
+				if err := os.Remove(tmpPath); err != nil {
+					rollbackErr = err
+				}
+			}
+		}
+		err = types.GenerateFunctionErrorWithRollback(err, rollbackErr)
+	}()
+
+	f, err := os.Create(tmpPath)
 	if err != nil {
 		return err
 	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(&obj); err != nil {
+		return err
+	}
 
-	return unix.Setxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue, 0)
+	if err := f.Close(); err != nil {
+		return err
+	}
+
+	return os.Rename(tmpPath, path)
 }
 
-func DeleteSnapshotHashInfoFromXattr(snapshotName string) error {
-	return unix.Removexattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName)
+func DeleteSnapshotHashInfoFromChecksumFile(snapshotName string) error {
```

Comment:
Rename to `DeleteSnapshotHashInfoChecksumFile`

Reply:
Updated, thank you.